File: projects/religious-voices/server/turnstile.py
"""Cloudflare Turnstile verification — Python port of the TS turnstile.ts."""

from __future__ import annotations

import os

import httpx

VERIFY_URL = "https://challenges.cloudflare.com/turnstile/v0/siteverify"


async def verify_turnstile(token: str | None, remote_ip: str | None) -> tuple[bool, str | None]:
    """Returns (ok, reason). Fails closed.

    In dev (TURNSTILE_SECRET_KEY unset and NODE_ENV != production),
    verification is skipped — local development is unblocked.
    """
    secret = os.environ.get("TURNSTILE_SECRET_KEY")
    if not secret:
        if os.environ.get("NODE_ENV") != "production":
            return True, "dev-skip"
        return False, "server-misconfigured"
    if not token:
        return False, "no-token"

    async with httpx.AsyncClient(timeout=10) as client:
        r = await client.post(
            VERIFY_URL,
            data={
                "secret": secret,
                "response": token,
                **({"remoteip": remote_ip} if remote_ip else {}),
            },
        )
        if r.status_code != 200:
            return False, f"siteverify-http-{r.status_code}"
        data = r.json()
    if not data.get("success"):
        return False, ",".join(data.get("error-codes", [])) or "denied"
    return True, None

File: src/lib/religious-voices/turnstile.ts
// Cloudflare Turnstile server-side verification.
//
// Fails open in dev (TURNSTILE_SECRET_KEY unset AND NODE_ENV !== production),
// fails closed in prod. The frontend gates submission on a token, so the
// only path here without a token is local dev.

const VERIFY_URL = "https://challenges.cloudflare.com/turnstile/v0/siteverify";

export async function verifyTurnstile(
  token: string | null | undefined,
  remoteIp: string | null,
): Promise<{ ok: boolean; reason?: string }> {
  const secret = process.env.TURNSTILE_SECRET_KEY;
  if (!secret) {
    if (process.env.NODE_ENV !== "production") return { ok: true, reason: "dev-skip" };
    return { ok: false, reason: "server-misconfigured" };
  }
  if (!token) return { ok: false, reason: "no-token" };

  const form = new URLSearchParams({ secret, response: token });
  if (remoteIp) form.set("remoteip", remoteIp);

  const r = await fetch(VERIFY_URL, {
    method: "POST",
    body: form,
    headers: { "Content-Type": "application/x-www-form-urlencoded" },
  });
  if (r.status !== 200) return { ok: false, reason: `siteverify-http-${r.status}` };
  const data = (await r.json()) as { success: boolean; "error-codes"?: string[] };
  if (!data.success) return { ok: false, reason: (data["error-codes"] ?? []).join(",") || "denied" };
  return { ok: true };
}
